dart/common/detail/AddonManager.h

```diff
@@ -0,0 +1,177 @@
+ *   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+ *   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+ *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef DART_COMMON_DETAIL_ADDONMANAGER_H_
+#define DART_COMMON_DETAIL_ADDONMANAGER_H_
+
+#include "dart/common/AddonManager.h"
+
+#define DART_COMMON_CHECK_ILLEGAL_ADDON_ERASE( Func, it, ReturnType )\
+  if(it ->second && !it ->second->isOptional(this))\
+  {\
+    dterr << "[AddonManager::" #Func << "] Illegal request to remove Addon!\n";\
+    assert(false);\
+    return ReturnType ;\
+  }
+
+namespace dart {
+namespace common {
+
+//==============================================================================
+template <class T>
+bool AddonManager::has() const
+{
+  return (get<T>() != nullptr);
+}
+
+//==============================================================================
+template <class T>
+T* AddonManager::get()
+{
+  AddonMap::iterator it = mAddonMap.find( typeid(T) );
+  if(mAddonMap.end() == it)
+    return nullptr;
+
+  return static_cast<T*>(it->second.get());
+}
+
+//==============================================================================
+template <class T>
+const T* AddonManager::get() const
+{
+  return const_cast<AddonManager*>(this)->get<T>();
+}
+
+//==============================================================================
+template <class T>
+void AddonManager::set(const T* addon)
+{
+  if(addon)
+  {
+    mAddonMap[typeid(T)] = addon->cloneAddon(this);
+    becomeManager(mAddonMap[typeid(T)].get(), false);
+  }
+  else
+  {
+    mAddonMap[typeid(T)] = nullptr;
+  }
+}
+
+//==============================================================================
+template <class T>
+void AddonManager::set(std::unique_ptr<T>&& addon)
+{
+  mAddonMap[typeid(T)] = std::move(addon);
+  becomeManager(mAddonMap[typeid(T)].get(), true);
+}
+
+//==============================================================================
+template <class T, typename ...Args>
+T* AddonManager::create(Args&&... args)
+{
+  T* addon = new T(this, std::forward<Args>(args)...);
+  mAddonMap[typeid(T)] = std::unique_ptr<T>(addon);
+  becomeManager(addon, false);
+
+  return addon;
+}
+
+//==============================================================================
+template <class T>
+void AddonManager::erase()
+{
+  AddonMap::iterator it = mAddonMap.find( typeid(T) );
+  DART_COMMON_CHECK_ILLEGAL_ADDON_ERASE(erase, it, DART_BLANK)
+  if(mAddonMap.end() != it)
+    it->second = nullptr;
```

Comment:
Yes, this is on purpose, and it's very important for two reasons:

1. It's possible for the user to access a specialized manager from a reference to the basic AddonManager. In that case, removing the entry entirely will definitely invalidate the iterator that's being held by the specialized manager.

2. All iterators for a ``std::map`` are guaranteed to remain valid as long as no entries are removed. Removing any entry from a ``std::map`` would risk invalidating any still existing iterators. So even if this entry is not a specialized addon, removing it from the map would risk invalidating any existing pointers to specialized entries.

It would probably be a good idea to put a developer's note about this in the comments of this function.